silo-core/contracts/incentives/base/BaseIncentivesController.sol

```diff
@@ -0,0 +1,198 @@
+// SPDX-License-Identifier: agpl-3.0
+pragma solidity 0.8.28;
+
+import {IERC20} from "openzeppelin5/token/ERC20/IERC20.sol";
+import {EnumerableSet} from "openzeppelin5/utils/structs/EnumerableSet.sol";
+
+import {DistributionTypes} from "../lib/DistributionTypes.sol";
+import {DistributionManager} from "./DistributionManager.sol";
+import {ISiloIncentivesController} from "../interfaces/ISiloIncentivesController.sol";
+
+/**
+ * @title BaseIncentivesController
+ * @notice Abstract contract template to build Distributors contracts for ERC20 rewards to protocol participants
+ * @author Aave
+  */
+abstract contract BaseIncentivesController is DistributionManager, ISiloIncentivesController {
+    using EnumerableSet for EnumerableSet.Bytes32Set;
+
+    mapping(address user => mapping(bytes32 programId => uint256 unclaimedRewards)) internal _usersUnclaimedRewards;
+
+    // this mapping allows whitelisted addresses to claim on behalf of others
+    // useful for contracts that hold tokens to be rewarded but don't have any native logic to claim Liquidity Mining
+    // rewards
+    mapping(address => address) internal _authorizedClaimers;
+
+    modifier onlyAuthorizedClaimers(address claimer, address user) {
+        if (_authorizedClaimers[user] != claimer) revert ClaimerUnauthorized();
+
+        _;
+    }
+
+    constructor(address _owner, address _notifier) DistributionManager(_owner, _notifier) {}
+
+    /// @inheritdoc ISiloIncentivesController
+    function createIncentivesProgram(DistributionTypes.IncentivesProgramCreationInput memory _incentivesProgramInput)
+        external
+        onlyOwner
+    {
+        require(bytes(_incentivesProgramInput.name).length > 0, InvalidIncentivesProgramName());
+
+        bytes32 programId = keccak256(abi.encodePacked(_incentivesProgramInput.name));
```

Comment:
I would not hash it - I would use bytes. Length of name will be limited but should be enough anyway. 
Working with hash will be harder, you always need to know original name and compare it eg when we scanning events, working experience will be poor.

Reply:
I added a name to the `IncentivesProgramCreated` event:
`event IncentivesProgramCreated(bytes32 indexed programId, string indexed name);`

For theGraph it should work well. We may have something like this:
```
type IncentivesProgram @entity(immutable: true) {
  id: Bytes!
  programId: Bytes!
  programName: String!
}

type RewardsAccrued @entity(immutable: true) {
 id: Bytes!
 program: IncentivesProgram!         <----- One-To-Many Relationships 
 user: Bytes!
 rewardToken: Bytes!
 amount: Int!
 ......
}
```

For scanning with custom scripts it is a matter of one extra call (which is not a big deal). You work with names and fetch an ID from the 'IncentivesProgramCreated' event.

Reply:
In addition I added fn:
`function getProgramId(string memory _programName) public pure returns (bytes32)`
This fn is used internally in the `SiloIncentivesController` and can be used from other smart contracts that will integrate with `SiloIncentivesController`.

Reply:
I would still vote for pure bytes without hashing it, this programs wil be managing by humans and it is easy to make mistake if you working with hashes and it will be pain in general.
and with pure bytes, you can simply use any combination of easily readable names with max UTF8 32 characters, it should be enough.
it will be much harder to make mistake that way

Reply:
Changed from the `keccak256(abi.encodePacked(_programName))` to the `bytes32(abi.encodePacked(_programName))`. Take a look.